apiserver/facades/client/resources/repository.go

```diff
@@ -0,0 +1,279 @@
+
+	// Origin holds the original source of a charm, including it's channel.
+	Origin corecharm.Origin
+
+	// Metadata is optional extra information about a particular model's
+	// "in-theatre" use use of the charm.
+	Metadata map[string]string
+}
+
+type CharmHub interface {
+	Refresh(ctx context.Context, config charmhub.RefreshConfig) ([]transport.RefreshResponse, error)
+}
+
+// TODO hml 2020-12-2
+// charmHubClient needs to be "caching" like the charmstoreclient is??
+type charmHubClient struct {
+	client CharmHub
+	id     CharmID
+}
+
+// ResolveResources, looks at the provided, charmhub and backend (already downloaded)
+// resources to determine which to use.  Provided (uploaded) take precedence.  If
+// charmhub has a newer resource than the back end, use that.
+// TODO: (hml) 2020-12-03
+// this logic will need refinement as work to incorporate charmhub resources continues.
+// Right now, just take the uploaded resources and add charmhub resources where resource
+// is missing
+func (ch *charmHubClient) ResolveResources(uploadedResources []charmresource.Resource) ([]charmresource.Resource, error) {
+	chResources, err := ch.listResources()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	for _, res := range uploadedResources {
+		chResources[res.Name] = res
+	}
+	results := make([]charmresource.Resource, len(chResources))
+	i := 0
+	for _, res := range chResources {
+		results[i] = res
+		i += 1
+	}
+	return results, nil
+}
+
+// listResources composes, a map of details for each of the charm's
+// resources. Those details are those associated with the specific
+// charm revision. They include the resource's metadata and revision.
+// Found via the CharmHub api.
+func (ch *charmHubClient) listResources() (map[string]charmresource.Resource, error) {
+	cfg, err := charmhub.DownloadOneFromChannel(ch.id.URL.Name, ch.id.Origin.Channel.String(), charmhub.RefreshPlatform(ch.id.Origin.Platform))
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	refreshResp, err := ch.client.Refresh(context.TODO(), cfg)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if len(refreshResp) == 0 {
+		return nil, errors.Errorf("no download refresh responses received")
+	}
+	resp := refreshResp[0]
+
+	if resp.Error != nil {
+		return nil, errors.Trace(errors.New(resp.Error.Message))
+	}
+	results := make(map[string]charmresource.Resource, len(resp.Entity.Resources))
+	for _, v := range resp.Entity.Resources {
+		var err error
+		results[v.Name], err = resourceFromRevision(v)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+	}
+	return results, nil
+}
+
+func resourceFromRevision(rev transport.ResourceRevision) (charmresource.Resource, error) {
+	resType, err := charmresource.ParseType(rev.Type)
+	if err != nil {
+		return charmresource.Resource{}, errors.Trace(err)
+	}
+	r := charmresource.Resource{
+		Meta: charmresource.Meta{
+			Name: rev.Name,
+			Type: resType,
+		},
+		Origin:   charmresource.OriginStore,
+		Revision: rev.Revision,
```

Comment:
In the qa example and unit tests, Revision and rev.Revision are both 0 at this point.  However by the time it gets backend.addPendingResources(), revision is -1.  Not sure why that is happening yet.